### PR TITLE
Update CircleCi cache key on build changes

### LINKED ISSUE #952: Strange test failure related to CircleCI caching
**Describe the bug**
I have had a whole slew of issues in #907 that I'm currently working on. I merged master to resolve some issues with the CI. This introduced some errors with CircleCi that have nothing to do with my PR. I think I finally tracked down the bug to be related to the caching in CircleCI.

Currently, we cache to the `test-{{ .Branch }}` key in `.circleci/config.yml`. In my case, the cache was not updated after the build system was changed by merging master (something to do with cartopy/numpy). This means all tests now fail, and no way to resolve the issue as the cache is only updated after a successful run.

Unfortunately, CircleCi has no way to clear the cache, the only way is to update the key.
I suggest we link the cache key to `setup.py` to avoid such issues in the future. This is also one of the solutions proposed in the CircleCI docs: https://support.circleci.com/hc/en-us/articles/115015426888-Clear-project-dependency-cache

I'm currently testing if this resolves the issue in #907.

### PR DESCRIPTION
**Pull request by stefsmeets**
## Description

I made this patch earlier today when I was debugging issues with CircleCI (#952). 

It turned out it did not actually solve my problem, but I think it makes caching in CircleCI a little bit safer in case we change the build through `setup.py` (see https://support.circleci.com/hc/en-us/articles/115015426888-Clear-project-dependency-cache). I linked it to `setup.py` since that is what CircleCI uses, but I'm open to other suggestions.

-   Closes #952

***

## Before you get started

-   [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## Checklist

-   [x] PR has a descriptive title for the [changelog](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#branches-pull-requests-and-code-review)
-   [x] Code follows the [style guide](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#code-style)
-   [x] [Documentation](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#documentation) is available for new functionality
-   [x] YAML files pass [`pre-commit`](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#pre-commit) or [`yamllint`](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/community/introduction.html#yaml) checks
-   [x] [Circle/CI tests pass](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#branches-pull-requests-and-code-review)
-   [x] [Codacy code quality checks pass](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#branches-pull-requests-and-code-review)
-   [x] [Documentation builds successfully](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#branches-pull-requests-and-code-review) on readthedocs
-   [x] [Unit tests](https://docs.esmvaltool.org/projects/esmvalcore/projects/esmvalcore/en/latest/contributing.html#contributing-to-the-esmvalcore-package) are available

***

To help with the number pull requests:

-   🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValTool/pulls) in this repository

### CODE REVIEW

**Comment by valeriupredoi** [comment]
@stefsmeets cheers for this! So as I understand it, now, the CI test cache will update (be re-created) everytime `setup.py` changes, right? Which is desirable, and I might add that probably best to concatenate `environment.yml` too - ?

**Comment by valeriupredoi** [comment]
something similar https://github.com/ESMValGroup/ESMValTool/issues/1998

File: .circleci/config.yml
Comment (valeriupredoi) [inline .circleci/config.yml:30]:
```suggestion
          key: test-{{ .Branch }}-{{ checksum "setup.py" }}-{{ checksum "environment.yml" }}
```

File: .circleci/config.yml
Comment (valeriupredoi) [inline .circleci/config.yml:38]:
```suggestion
          key: test-{{ .Branch }}-{{ checksum "setup.py" }}-{{ checksum "environment.yml" }}
```

File: .circleci/config.yml
Comment (valeriupredoi) [inline .circleci/config.yml:56]:
```suggestion
          key: install-{{ .Branch }}-{{ checksum "setup.py" }}-{{ checksum "environment.yml" }}
```

File: .circleci/config.yml
Comment (valeriupredoi) [inline .circleci/config.yml:76]:
```suggestion
          key: install-{{ .Branch }}-{{ checksum "setup.py" }}-{{ checksum "environment.yml" }}
```

**Review by valeriupredoi — CHANGES_REQUESTED** [review]
here, I've added the extra check on `environment.yml` too, how's it sound?

**Comment by stefsmeets** [comment]
> @stefsmeets cheers for this! So as I understand it, now, the CI test cache will update (be re-created) everytime `setup.py` changes, right? Which is desirable, and I might add that probably best to concatenate `environment.yml` too - ?

Yeah, that is the idea.I thought that `environment.yml` is not necessary, since we pull in the docker container where these changes in `environment.yml` should proliferate, but maybe it does.

I think this means we can remove the branch too, but I'm a bit hesitant. In that way the cache will be shared between branched I think, which may be nice.

**Comment by valeriupredoi** [comment]
yeah, errm I'm not too sure about `Branch` either, I'll let @bouweandela comment on that. Cheers for the inclusion of my suggestions, better safe than sorry! Gonna approve but I'd think @bouweandela should have a look too before we merge, he's written that code anyways :beer:

**Review by valeriupredoi — APPROVED** [review]

**Review by bouweandela — CHANGES_REQUESTED** [review]
I do not think these changes are needed, because if you look at the paths that are cached, you see that we only cache the package cache of pip and conda across runs in the same branch, not the actual environment that is created based on the list of dependencies in setup.py/environment.yml/etc.

**Comment by stefsmeets** [comment]
> I do not think these changes are needed, because if you look at the paths that are cached, you see that we only cache the package cache of pip and conda across runs in the same branch, not the actual environment that is created based on the list of dependencies in setup.py/environment.yml/etc.

It's not about the environment, but the pip cache is exactly what may give unexpected results. 

The issue I ran into, which is why I think this is useful, is when I had to make an update to `esmvaltool_sample_data`. I had to update the version number in `setup.py` of `esmvaltool_sample_data`, just to see the change back in CircleCi. It meant I could not just point to the commit of the branch to test if it works. It took me a while to figure out and made it difficult to test.

**Comment by bouweandela** [comment]
> It meant I could not just point to the commit of the branch to test if it works. It actually downloaded the correct version, but installed the cached one (???)

That sounds like a bug in pip? Maybe you could raise the issue on their issue tracker? https://github.com/pypa/pip/issues

**Comment by stefsmeets** [comment]
> 
> 
> > It meant I could not just point to the commit of the branch to test if it works. It actually downloaded the correct version, but installed the cached one (???)
> 
> That sounds like a bug in pip? Maybe you could raise the issue on their issue tracker? https://github.com/pypa/pip/issues

If you don't think this PR is useful, feel free to close the PR. I don't have a handle on whether this is a bug with CircleCI or pip, or something circumstantial. This PR will avoid the issue I was having with our build system.

**Comment by valeriupredoi** [comment]
I reckon, if this is not affecting the functionality of the tests on Circle, and it solves @stefsmeets problem, it's good to merge - it'd be better to get to the bottom of the problem so it won't bite us in the back in future, in a different shape, but for now I am pro merge :+1:

**Comment by bouweandela** [comment]
Thanks @stefsmeets! I'll close the pull request then, because I think it's important that our tests are as fast as reasonably possible and this looks like it's a bug in `pip`. These caches exist on everyone's computer, so the caching is not specific to CircleCI.

**Pull request closed without merging**
